nowcasting_api/cache.py

```diff
@@ -1,208 +1,112 @@
-    :param response: dict of responses, same keys as last_updated
-    :param remove_cache_time_seconds: the amount of time, after which the cache should be removed
+
+def generate_cache_key(func: Callable, *args, **kwargs) -> str:
+    """Generate a unique cache key based on the endpoint path and query parameters.
+
+    :param func: The route handler function
+    :param request: The FastAPI request object
+    :return: A string to be used as cache key
     """
-    now = datetime.now(tz=timezone.utc)
-    logger.info("Checking and removing old cache entries")
-    keys_to_remove = []
-    last_updated_copy = last_updated.copy()
-    for key, value in last_updated_copy.items():
-        if now - timedelta(seconds=remove_cache_time_seconds) > value:
-            logger.debug(f"Removing {key} from cache, ({value})")
-            keys_to_remove.append(key)
-
-    del last_updated_copy
-    logger.debug(f"Removing {len(keys_to_remove)} keys from cache")
-
-    for key in keys_to_remove:
-        try:
-            last_updated.pop(key)
-            response.pop(key)
-        except KeyError:
-            logger.warning(
-                f"Could not remove {key} from cache. "
-                f"This could be because it has already been removed"
-            )
-
-    return last_updated, response
-
-
-def cache_response(func):
+    # Create key from path and sorted query params
+    request: Request = kwargs.get("request")
+    path = request.url.path if request else "unknown"
+    query_params = sorted(request.query_params.items()) if request else []
+    key = f"{path}:{query_params}"
+    logger.debug(f"Generated cache key: {key}")
+    return key
+
+
+def save_api_call(
+    request: Request = None, user: Optional[Any] = None, session: Optional[Any] = None
+):
+    """
+    Save API call to database.
+
+    This can be used as a dependency in FastAPI routes.
+
+    :param request: The FastAPI request object
+    :param user: The user making the request, if authenticated
+    :param session: The database session
     """
-    Decorator that caches the response of a FastAPI async function.
+    save_api_call_to_db(session=session, user=user, request=request)
+    return request
+
+
+def clear_cache_key(key: str, expiration: int = DELETE_CACHE_TIME_SECONDS):
+    """Clear a specific cache key from the FastAPI cache.
 
     Example:
     ```
-        app = FastAPI()
+    @app.delete("/clear-cache/{item_id}")
+    async def clear_item_cache(item_id: str):
+        key = f"/items/{item_id}"
+        clear_cache_key(key)
+        return {"message": f"Cache cleared for item {item_id}"}
+    ```
+
+    :param key: The cache key to clear
+    :param expiration: Time in seconds before the key can be cached again
+    """
+    try:
+        FastAPICache.get_backend().clear(namespace="api", key=key)
+        logger.info(f"Cleared cache key: {key}")
+    except Exception as e:
+        logger.error(f"Failed to clear cache for key {key}: {e}")
+
```

Comment:
how does this function use `expiration`?

Reply:
I apologize for my oversight in not pushing the recent changes.

Reply:
It might be worth removing `DELETE_CACHE_TIME_SECONDS` and `expiration`, as right now it doesnt make sense. 

where does `clear_cache_key` get used? do we need this? does the library do this anyway?

Reply:
The expiration parameter in `clear_cache_key` is used to set how long the cache key should be locked before it can be cached again. This creates a lock by setting an entry with the key format `{key}:lock` that expires after the specified time (by default, it uses `DELETE_CACHE_TIME_SECONDS=240`).